test/lit/passes/poppify.wast

```diff
@@ -0,0 +1,411 @@
+  (func $try-catch (result i32)
+    (try i32
+      (do
+        (throw $e
+          (i32.const 0)
+        )
+      )
+      (catch $e
+        (return
+          (pop i32)
+        )
+      )
+      (catch_all
+        (i32.const 1)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $tuple (result i32 i64)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (i64.const 1)
+  ;; CHECK-NEXT: )
+  (func $tuple (result i32 i64)
+    (tuple.make
+      (i32.const 0)
+      (i64.const 1)
+    )
+  )
+
+  ;; CHECK:      (func $extract-first (result i32)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (i64.const 1)
+  ;; CHECK-NEXT:  (f32.const 2)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop f32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i64)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $extract-first (result i32)
+    (tuple.extract 0
+      (tuple.make
+        (i32.const 0)
+        (i64.const 1)
+        (f32.const 2)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $extract-middle (result i64)
+  ;; CHECK-NEXT:  (local $0 i64)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (i64.const 1)
+  ;; CHECK-NEXT:  (f32.const 2)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop f32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $0
+  ;; CHECK-NEXT:   (pop i64)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $0)
+  ;; CHECK-NEXT: )
+  (func $extract-middle (result i64)
+    (tuple.extract 1
+      (tuple.make
+        (i32.const 0)
+        (i64.const 1)
+        (f32.const 2)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $extract-last (result f32)
+  ;; CHECK-NEXT:  (local $0 f32)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (i64.const 1)
+  ;; CHECK-NEXT:  (f32.const 2)
+  ;; CHECK-NEXT:  (local.set $0
+  ;; CHECK-NEXT:   (pop f32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i64)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $0)
+  ;; CHECK-NEXT: )
+  (func $extract-last (result f32)
+    (tuple.extract 2
+      (tuple.make
+        (i32.const 0)
+        (i64.const 1)
+        (f32.const 2)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $drop
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $drop
+    (drop
+      (i32.const 0)
+    )
+  )
+
+  ;; CHECK:      (func $drop-tuple
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT:  (i64.const 1)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i64)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (pop i32)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $drop-tuple
+    (drop
+      (tuple.make
+        (i32.const 0)
+        (i64.const 1)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $local-get-tuple (result i32 i64)
+  ;; CHECK-NEXT:  (local $x (i32 i64))
```

Comment:
When splitting a global, we seem to remove the original tuple global. Would it be good to do that for locals too? (Not necessarily in this PR though)

Reply:
I plan to use existing passes like `CoalesceLocals` to do that as a follow-up optimization. I guess in principle we could use `RemoveUnusedModuleElements` to clean up the globals as well, but that's simple enough to do here.